packages/amplify-graphql-relational-transformer/src/schema.ts

```diff
@@ -148,15 +149,18 @@ export function ensureHasOneConnectionField(
   const filterInputName = toPascalCase(['Model', object.name.value, 'FilterInput']);
   const filterInput = ctx.output.getType(filterInputName) as InputObjectTypeDefinitionNode;
   if (filterInput) {
-    ctx.output.putType(updateFilterConnectionInputWithConnectionField(filterInput, connectionAttributeName));
+    // use the mapped connection attribute name for FilterInput because the input can be nested which is a mess to map in VTL
+    // instead the input will retain the original field name
+    ctx.output.putType(updateFilterConnectionInputWithConnectionField(filterInput, mappedConnectionAttributeName));
   }
 
   const conditionInputName = toPascalCase(['Model', object.name.value, 'ConditionInput']);
   const conditionInput = ctx.output.getType(conditionInputName) as InputObjectTypeDefinitionNode;
   if (conditionInput) {
-    ctx.output.putType(updateFilterConnectionInputWithConnectionField(conditionInput, connectionAttributeName));
+    // use the mapped connection attribute name for ConditionInput because the input can be nested which is a mess to map in VTL
+    // instead the input will retain the original field name
```

Comment:
The issue is that the list filter input is recursive.

Consider a schema like:
```graphql
type Todo @model @mapsTo(name: "OrigTodo") {
  id: ID!
  comments: [Comment] @hasMany
}

type Comment @model @mapsTo(name: "OrigComment") {
  id: ID!
  todo: Todo @belongsTo
}
```
Without this line of code, the generated `ModelCommentFilterInput` would be:
```graphql
input ModelCommentFilterInput {
  id: ModelIDInput
  and: [ModelCommentFilterInput]
  or: [ModelCommentFilterInput]
  not: ModelCommentFilterInput
  todoCommentsId: ModelIDInput # <== this is the problem
}
```
But the "OrigTodo" table contains entries with "origTodoCommentsId" as the foreign key field. Because the filter input is recursive, this cannot be mapped in VTL because [VTL does not support recursion or while loops](http://people.apache.org/~henning/velocity/html/ch05s04.html#:~:text=Unlike%20other%20languages%2C%20the%20Velocity%20templating%20language%20allows%20only%20loops%20where%20the%20number%20of%20iterations%20is%20predetermined.%20There%20is%20no%20do..while%20or%20repeat..until%20loop%20in%20Velocity). So instead we generate the FilterInput type with the original foreign key field name:
```graphql
input ModelCommentFilterInput {
  id: ModelIDInput
  and: [ModelCommentFilterInput]
  or: [ModelCommentFilterInput]
  not: ModelCommentFilterInput
  origTodoCommentsId: ModelIDInput # <== this fixes it
}
```

Reply:
Would this have any impact on model generation? IIRC the filers were added specifically for Data Store, so could we just validate if this works in all the libs

Reply:
This has now been fixed. The FilterInput and ConditionInput is mapped in a lambda function so no impact to the generated API anymore